In [1]:
# imports
import json
import pandas as pd

# EDA - User Data

In [2]:
# import json data
users = []
for line in open('users.json', 'r'):
    users.append(json.loads(line))
    
# put data into dataFrame
users = pd.DataFrame(users)

In [3]:
# check head of dataFrame
users.head()

_id  active               createdDate  \
0  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   

                  lastLogin      role signUpSource state  
0  {'$date': 1609687537858}  consumer        Email    WI  
1  {'$date': 1609687537858}  consumer        Email    WI  
2  {'$date': 1609687537858}  consumer        Email    WI  
3  {'$date': 1609687530597}  consumer        Email    WI  
4  {'$date': 1609687537858}  consumer        Email    WI

In [4]:
# summary of dataFrame

users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   _id           495 non-null    object
 1   active        495 non-null    bool  
 2   createdDate   495 non-null    object
 3   lastLogin     433 non-null    object
 4   role          495 non-null    object
 5   signUpSource  447 non-null    object
 6   state         439 non-null    object
dtypes: bool(1), object(6)
memory usage: 23.8+ KB


### Takeaways

    1. There are 495 user IDs provided. Are all unique or any duplicates?
    2. There is an extra column (signUpSource) that was not provided in the data schema.

## EDA of _id column

In [5]:
# view column names of dataframe
users.columns

Index(['_id', 'active', 'createdDate', 'lastLogin', 'role', 'signUpSource',
       'state'],
      dtype='object')

In [6]:
# Return the rows with duplicate user IDs and put in new dataframe
dupl_users = users.loc[users['_id'].duplicated(),:]

# view dataframe
dupl_users

_id  active               createdDate  \
1    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
4    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
5    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
8    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
..                                    ...     ...                       ...   
490  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
491  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
492  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
493  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
494  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   

                    lastLogin         role signUpSource state  
1    {'$date': 1609687537858}     consumer        Email    WI  
2    {'$date': 1609687537858}     consumer        Email    WI  
4    {'$date': 1609687537858}     consumer        Email    WI  
5    {'$date': 1609687537858}     consumer        Email    WI  
8    {'$date': 1609687537858}     consumer        Email    WI  
..                        ...          ...          ...   ...  
490  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
491  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
492  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
493  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
494  {'$date': 1614963143204}  fetch-staff          NaN   NaN  

[283 rows x 7 columns]

In [33]:
# view first 50 users in Duplicate Users dataframe

dupl_users.head(50)

_id  active               createdDate  \
1   {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2   {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
4   {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
5   {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
8   {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
10  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
11  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
12  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
13  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
14  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
15  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
18  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
19  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
24  {'$oid': '5ff370c562fde912123a5e0e'}    True  {'$date': 1609789637776}   
26  {'$oid': '5ff36d0362fde912123a5535'}    True  {'$date': 1609788675973}   
27  {'$oid': '5ff36be7135e7011bcb856d3'}    True  {'$date': 1609788391239}   
29  {'$oid': '5ff36d0362fde912123a5535'}    True  {'$date': 1609788675973}   
31  {'$oid': '5ff36a3862fde912123a4460'}    True  {'$date': 1609787960352}   
32  {'$oid': '5ff370c562fde912123a5e0e'}    True  {'$date': 1609789637776}   
33  {'$oid': '5ff36a3862fde912123a4460'}    True  {'$date': 1609787960352}   
34  {'$oid': '5ff36a3862fde912123a4460'}    True  {'$date': 1609787960352}   
35  {'$oid': '5ff370c562fde912123a5e0e'}    True  {'$date': 1609789637776}   
36  {'$oid': '5ff36a3862fde912123a4460'}    True  {'$date': 1609787960352}   
37  {'$oid': '5ff370c562fde912123a5e0e'}    True  {'$date': 1609789637776}   
39  {'$oid': '5ff36be7135e7011bcb856d3'}    True  {'$date': 1609788391239}   
43  {'$oid': '5ff47392c3d63511e2a47881'}    True  {'$date': 1609855890233}   
44  {'$oid': '5ff4ce33c3d63511e2a484b6'}    True  {'$date': 1609879091229}   
46  {'$oid': '5ff47392c3d63511e2a47881'}    True  {'$date': 1609855890233}   
48  {'$oid': '5ff4ce33c3d63511e2a484b6'}    True  {'$date': 1609879091229}   
49  {'$oid': '5ff4ce3dc3d63511e2a484dc'}    True  {'$date': 1609879101887}   
50  {'$oid': '5ff4ce33c3d63511e2a484b6'}    True  {'$date': 1609879091229}   
51  {'$oid': '5ff47392c3d63511e2a47881'}    True  {'$date': 1609855890233}   
52  {'$oid': '5ff4ce3dc3d63511e2a484dc'}    True  {'$date': 1609879101887}   
53  {'$oid': '5ff4ce3dc3d63511e2a484dc'}    True  {'$date': 1609879101887}   
54  {'$oid': '5ff4ce33c3d63511e2a484b6'}    True  {'$date': 1609879091229}   
56  {'$oid': '5ff4ce33c3d63511e2a484b6'}    True  {'$date': 1609879091229}   
57  {'$oid': '5ff4ce33c3d63511e2a484b6'}    True  {'$date': 1609879091229}   
59  {'$oid': '5ff47392c3d63511e2a47881'}    True  {'$date': 1609855890233}   
61  {'$oid': '5ff5d15aeb7c7d12096d91a2'}    True  {'$date': 1609945434680}   
62  {'$oid': '5ff5d15aeb7c7d12096d91a2'}    True  {'$date': 1609945434680}   
63  {'$oid': '5ff5d15aeb7c7d12096d91a2'}    True  {'$date': 1609945434680}   
64  {'$oid': '5ff5d15aeb7c7d12096d91a2'}    True  {'$date': 1609945434680}   
65  {'$oid': '5ff5d15aeb7c7d12096d91a2'}    True  {'$date': 1609945434680}   
67  {'$oid': '5ff5d15aeb7c7d12096d91a2'}    True  {'$date': 1609945434680}   
68  {'$oid': '5ff5d15aeb7c7d12096d91a2'}    True  {'$date': 1609945434680}   
69  {'$oid': '5ff5d15aeb7c7d12096d91a2'}    True  {'$date': 1609945434680}   
70  {'$oid': '5ff5d15aeb7c7d12096d91a2'}    True  {'$date': 1609945434680}   
71  {'$oid': '5ff5d15aeb7c7d12096d91a2'}    True  {'$date': 1609945434680}   
72  {'$oid': '5ff5d15aeb7c7d12096d91a2'}    True  {'$date': 1609945434680}   
73  {'$oid': '5ff5d15aeb7c7d12096d91a2'}    True  {'$date': 1609945434680}   

                   lastLogin      role signUpSource state 

In [32]:
# view last 50 users in Duplicate Users dataframe

dupl_users.tail(50)

_id  active               createdDate  \
441  {'$oid': '5fc961c3b8cfca11a077dd33'}    True  {'$date': 1607033283936}   
442  {'$oid': '5fc961c3b8cfca11a077dd33'}    True  {'$date': 1607033283936}   
443  {'$oid': '5fc961c3b8cfca11a077dd33'}    True  {'$date': 1607033283936}   
444  {'$oid': '5fc961c3b8cfca11a077dd33'}    True  {'$date': 1607033283936}   
445  {'$oid': '5fc961c3b8cfca11a077dd33'}    True  {'$date': 1607033283936}   
446  {'$oid': '5fc961c3b8cfca11a077dd33'}    True  {'$date': 1607033283936}   
447  {'$oid': '5fc961c3b8cfca11a077dd33'}    True  {'$date': 1607033283936}   
448  {'$oid': '5fc961c3b8cfca11a077dd33'}    True  {'$date': 1607033283936}   
449  {'$oid': '5fc961c3b8cfca11a077dd33'}    True  {'$date': 1607033283936}   
450  {'$oid': '5fc961c3b8cfca11a077dd33'}    True  {'$date': 1607033283936}   
451  {'$oid': '5fc961c3b8cfca11a077dd33'}    True  {'$date': 1607033283936}   
452  {'$oid': '5fc961c3b8cfca11a077dd33'}    True  {'$date': 1607033283936}   
453  {'$oid': '5fc961c3b8cfca11a077dd33'}    True  {'$date': 1607033283936}   
454  {'$oid': '5fc961c3b8cfca11a077dd33'}    True  {'$date': 1607033283936}   
457  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
458  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
459  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
460  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
461  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
463  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
464  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
465  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
466  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
467  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
468  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
469  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
470  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
471  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
472  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
473  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
474  {'$oid': '5fa41775898c7a11a6bcef3e'}    True  {'$date': 1604589429396}   
476  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
477  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
478  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
479  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
480  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
481  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
482  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
483  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
484  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
485  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
486  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
487  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
488  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
489  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
490  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
491  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
492  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
493  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
494  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   

        

In [8]:
# Looks like there is more than one role. how many unique roles?

dupl_users['role'].nunique()

2

### User Data Takeaways
1. There are duplicate user IDs (_id column) in the users data. 
    
    _id is the primary key and values in this column should be unique. Looks like the duplicate user IDs have the same creation date. Maybe they were all created by accident on the same day by the same user? 
     
     
2. There is an additional role ('fetch-staff') in the role column that wasn't mentioned in User data schema. 
    
    Some of the duplicate user IDs were created by fetch-staff. Maybe test accounts?
    
    
3. There is an additional column ('signUpSource' column) that was not mentioned in the User Data Schema. 
   
    Should this column be added to the schema or removed from the data? 